File: app/src/main/java/com/vwoom/timelapsecollection/adapters/DetailsAdapter.java
package com.vwoom.timelapsecollection.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.vwoom.timelapsecollection.R;
import com.vwoom.timelapsecollection.database.entry.PhotoEntry;

import java.io.File;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DetailsAdapter extends RecyclerView.Adapter<DetailsAdapter.DetailsAdapterViewHolder> {

    private final static String TAG = DetailsAdapter.class.getSimpleName();

    private List<PhotoEntry> mPhotos;
    private final DetailsAdapterOnClickHandler mClickHandler;

    public interface DetailsAdapterOnClickHandler {
        void onClick(PhotoEntry clickedPhoto);
    }

    public DetailsAdapter(DetailsAdapterOnClickHandler handler){ mClickHandler = handler;}

    public class DetailsAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.detail_thumbnail) ImageView mDetailThumbnail;

        public DetailsAdapterViewHolder(View view){
            super(view);
            ButterKnife.bind(this, view);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            PhotoEntry clickedPhoto = mPhotos.get(adapterPosition);
            mClickHandler.onClick(clickedPhoto);
        }
    }

    @NonNull
    @Override
    public DetailsAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        int layoutIdForGridItem = R.layout.detail_recyclerview_item;
        LayoutInflater inflater = LayoutInflater.from(context);
        boolean shouldAttachToParentImmediately = false;
        View view = inflater.inflate(layoutIdForGridItem, parent, shouldAttachToParentImmediately);
        return new DetailsAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull DetailsAdapterViewHolder holder, int position) {
        PhotoEntry currentPhoto = mPhotos.get(position);
        String photo_path = currentPhoto.getUrl();
        File f = new File(photo_path);
        Glide.with(holder.itemView.getContext())
                .load(f)
                .into(holder.mDetailThumbnail);
    }

    @Override
    public int getItemCount() {
        if (mPhotos == null) return 0;
        return mPhotos.size();
    }

    public void setPhotoData(List<PhotoEntry> photoData){
        mPhotos = photoData;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/vwoom/timelapsegallery/details/DetailsAdapter.kt
package com.vwoom.timelapsegallery.details

import android.content.Context
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.vwoom.timelapsegallery.R
import com.vwoom.timelapsegallery.details.DetailsAdapter.DetailsAdapterViewHolder
import com.vwoom.timelapsegallery.data.entry.PhotoEntry
import com.vwoom.timelapsegallery.data.view.Project
import com.vwoom.timelapsegallery.databinding.DetailRecyclerviewItemBinding
import com.vwoom.timelapsegallery.databinding.GalleryRecyclerviewItemBinding
import com.vwoom.timelapsegallery.utils.FileUtils
import kotlinx.android.synthetic.main.detail_recyclerview_item.view.*
import java.io.File

class DetailsAdapter(private val mClickHandler: DetailsAdapterOnClickHandler, context: Context) : RecyclerView.Adapter<DetailsAdapterViewHolder>() {
    private var mPhotos: List<PhotoEntry>? = null
    private var mProject: Project? = null
    private var mCurrentPhoto: PhotoEntry? = null
    private val mExternalFilesDir: File?

    interface DetailsAdapterOnClickHandler {
        fun onClick(clickedPhoto: PhotoEntry)
    }

    inner class DetailsAdapterViewHolder(var binding: DetailRecyclerviewItemBinding) : RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        override fun onClick(view: View) {
            val adapterPosition = adapterPosition
            val clickedPhoto = mPhotos!![adapterPosition]
            mCurrentPhoto = clickedPhoto
            mClickHandler.onClick(clickedPhoto)
        }

        init {
            binding.root.setOnClickListener(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DetailsAdapterViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val shouldAttachToParentImmediately = false
        val binding = DetailRecyclerviewItemBinding.inflate(inflater, parent, shouldAttachToParentImmediately)
        return DetailsAdapterViewHolder(binding)
    }

    override fun onBindViewHolder(holder: DetailsAdapterViewHolder, position: Int) {
        val binding = holder.binding
        val context = holder.itemView.context
        val currentPhoto = mPhotos?.get(position)
        val photo_path = FileUtils.getPhotoUrl(mExternalFilesDir!!, mProject!!, currentPhoto!!)
        val f = File(photo_path)
        // TODO (update) dynamically resize detail view
        Glide.with(context)
                .load(f)
                .centerCrop()
                .into(binding.detailThumbnail)
        if (mPhotos!![position] === mCurrentPhoto) {
            binding.selectionIndicator.visibility = View.VISIBLE
        } else {
            binding.selectionIndicator.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        return if (mPhotos == null) 0 else mPhotos!!.size
    }

    fun setPhotoData(photoData: List<PhotoEntry>?, project: Project?) {
        mPhotos = photoData
        mProject = project
        notifyDataSetChanged()
    }

    fun setCurrentPhoto(photo: PhotoEntry?) {
        mCurrentPhoto = photo
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = DetailsAdapter::class.java.simpleName
    }

    init {
        mExternalFilesDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
    }
}